### ISSUE TITLE
Feature request: Weighted average for EmbeddingBag

### ISSUE THREAD

**Issue by kunaldahiya** [pytorch/pytorch#4068]
Right now 'torch.nn.EmbeddingBag' supports only 'sum' and 'mean'. What do you think about providing an option for weights to compute 'weighted average'? This would be more memory efficient than using current alternative. 

For instance something like 'sp_weights' in 'tf.nn.embedding_lookup_sparse' [1].

References:
[1] https://www.tensorflow.org/api_docs/python/tf/nn/embedding_lookup_sparse

**Comment by glample** [comment]
Hi, any updates on this? Being able to provide weights (on top of the indices) would be really useful.

**Comment by snakers4** [comment]
This would be very helpful for my work with the Russian language.
I understand that I can use some kind of attention + small char level CNN, but I really doubt that my PyTorch implementation will not be 10x slower.
Many thanks!

**Comment by snakers4** [comment]
As some kind of motivation, I will just post a [link](https://towardsdatascience.com/complexity-generalization-computational-cost-in-nlp-modeling-of-morphologically-rich-languages-7fa2c0b45909) to my post, where in many applications for Russian EmbeddingBags were superior to BPE =)

**Comment by zou3519** [comment]
API Bikeshedding: which of these two APIs would be better?

1) nn.EmbeddingBag's forward pass accepts a `per_input_weights` argument. When `mode='sum'`, this makes it do a weighted sum; when `mode='mean'`, this does a weighted mean, when `mode='max'`, this does a weighted max, like the TF API.

2) nn.EmbeddingBag's forward pass accepts a `per_input_weights` argument and a new `mode='weighted_sum'`. `mode='weighted_sum'` weights the output of the embedding according to the weights. No weighted mean / weighted max are implemented.

I'm leaning towards (2) because I haven't been able to find use cases for "weighted mean" (which can be emulated via weighted sum) and "weighted max".

**Comment by ezyang** [comment]
FWIW, you don't have to actually implemented weighted mean and weighted max if you implement (1); you can just make them raise errors. (This is not necessarily in favor of (1), but it's a comment on the reasoning.)

**Comment by snakers4** [comment]
> nn.EmbeddingBag's forward pass accepts a per_input_weights argument

Most likely these weights will be calculated using some sort of attention mechanism
The simplest attention mechanism would be something like [this](https://gist.github.com/cbaziotis/94e53bdd6e4852756e0395560ff38aa4) (just a linear layer + softmax)

I wonder whether something like this can be implemented inside of this layer

**Comment by zou3519** [comment]
Added the feature in #18957.

**Comment by snakers4** [comment]
Many thanks!
We will try adding some this to our next project!

**Comment by drevicko** [comment]
Is there currently a plan to implement per_sample_weights on CUDA for max aggregation?
